scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamFunctions.scala

```diff
@@ -58,4 +62,35 @@ object BeamFunctions {
       }
     }
   }
+
+  case class HashJoinFn[K, V, U, W](
+    joiner: (K, V, Iterable[U]) => Iterator[W],
+    sideInput: PCollectionView[java.util.Map[K, java.lang.Iterable[U]]]
+  ) extends DoFn[(K, V), (K, W)] {
+    private[this] var mapRight: java.util.Map[K, java.lang.Iterable[U]] = null
+    private[this] val emptyUs: Iterable[U] = Seq.empty[U]
+
+    @ProcessElement
+    def processElement(c: DoFn[(K, V), (K, W)]#ProcessContext): Unit = {
+      if (mapRight == null) {
+        mapRight = c.sideInput(sideInput)
+      }
+      val key = c.element()._1
```

Comment:
nit: `val (kev, value) = c.element()`

Reply:
I may have suggested this. Actually here where we are in the core, I'm actually pretty supportive of coding it like Java.

Reducing a few allocations inside a big job could be measurable.